src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,397 @@
+        ←adjugate_pow],
+    cases le_total m n with hm hm,
+    { obtain ⟨k, rfl⟩ := le_iff_exists_add.mp hm,
+      rw [pow_add, mul_eq_mul, ←matrix.mul_assoc, adjugate_mul, ←mul_eq_mul (A ^ m), ←pow_add,
+          add_comm, pow_add, mul_eq_mul (A ^ k), matrix.mul_assoc, mul_adjugate, mul_smul, smul_mul,
+          matrix.mul_one, matrix.one_mul] },
+    { obtain ⟨k, rfl⟩ := le_iff_exists_add.mp hm,
+      rw [pow_add, mul_eq_mul, adjugate_mul_distrib, matrix.mul_assoc, ←adjugate_mul_distrib,
+          adjugate_mul, mul_smul, matrix.mul_one, ←mul_eq_mul _ (A ^ k), ←pow_add,
+          add_comm, pow_add, mul_eq_mul (A ^ k), adjugate_mul_distrib, ←matrix.mul_assoc,
+          mul_adjugate, smul_mul, matrix.one_mul] } },
+  { rw [A.nonsing_inv_apply_not_is_unit h],
+    cases m,
+    { rw [pow_zero, matrix.mul_one, matrix.one_mul] },
+    { rw [zero_pow (nat.zero_lt_succ m), matrix.mul_zero, matrix.zero_mul] } }
+end
+
+end nat_pow
+
+section int_pow
+open int
+
+@[simp] theorem one_fpow : ∀ (n : ℤ), (1 : M) ^ n = 1
+| (n : ℕ) := by rw [gpow_coe_nat, one_pow]
+| -[1+ n] := by rw [gpow_neg_succ_of_nat, one_pow, inv_one]
+
+lemma zero_fpow : ∀ z : ℤ, z ≠ 0 → (0 : M) ^ z = 0
+| (n : ℕ) h := by { rw [gpow_coe_nat, zero_pow], refine lt_of_le_of_ne n.zero_le (ne.symm _),
+  simpa using h  }
+| -[1+n]  h := by simp [zero_pow n.zero_lt_succ]
+
+lemma fzero_pow_eq (n : ℤ) : (0 : M) ^ n = if n = 0 then 1 else 0 :=
+begin
+  split_ifs with h,
+  { rw [h, gpow_zero] },
+  { rw [zero_fpow _ h] }
+end
+
+theorem inv_fpow (A : M) : ∀n:ℤ, A⁻¹ ^ n = (A ^ n)⁻¹
+| (n : ℕ) := by rw [gpow_coe_nat, gpow_coe_nat, inv_pow']
+| -[1+ n] := by rw [gpow_neg_succ_of_nat, gpow_neg_succ_of_nat, inv_pow']
+
+@[simp] lemma fpow_neg_one (A : M) : A ^ (-1 : ℤ) = A⁻¹ :=
+begin
+  convert div_inv_monoid.gpow_neg' 0 A,
+  simp only [gpow_one, int.coe_nat_zero, int.coe_nat_succ, gpow_eq_pow, zero_add]
+end
+
+theorem fpow_coe_nat (A : M) (n : ℕ) : A ^ (n : ℤ) = (A ^ n) :=
+gpow_coe_nat _ _
+
+@[simp] theorem fpow_neg_coe_nat (A : M) (n : ℕ) : A ^ (-n : ℤ) = (A ^ n)⁻¹ :=
+begin
+  cases n,
+  { simp },
+  { exact div_inv_monoid.gpow_neg' _ _ }
+end
+
+
+lemma _root_.is_unit.det_fpow {A : M} (h : is_unit A.det) (n : ℤ) : is_unit (A ^ n).det :=
+begin
+  cases n,
+  { simpa using h.pow n },
+  { simpa using h.pow n.succ }
+end
+
+lemma is_unit_det_fpow_iff {A : M} {z : ℤ} :
+  is_unit (A ^ z).det ↔ is_unit A.det ∨ z = 0 :=
+begin
+  induction z using int.induction_on with z IH z IH,
+  { simp },
+  { rw [←int.coe_nat_succ, fpow_coe_nat, det_pow, is_unit_pos_pow_iff (z.zero_lt_succ),
+        ←int.coe_nat_zero, int.coe_nat_eq_coe_nat_iff],
+    simp },
+  { rw [←neg_add', ←int.coe_nat_succ, fpow_neg_coe_nat, is_unit_nonsing_inv_det_iff,
+        det_pow, is_unit_pos_pow_iff (z.zero_lt_succ), neg_eq_zero, ←int.coe_nat_zero,
+        int.coe_nat_eq_coe_nat_iff],
+    simp }
+end
+
+lemma inv_fpow' {a : M} (h : is_unit a.det) (n : ℤ) :
```

Comment:
Starting from this point you start to mix `a` and `A`. I would prefer `A` throughout.
```suggestion
lemma inv_fpow' {A : M} (h : is_unit A.det) (n : ℤ) :
```